US-Trustee-Program/Bankruptcy-Oversight-Support-Systems
# CAMS-81 Initial setup of CosmoDb

# Purpose

Azure CosmoDb is the chosen data storage solution for CAMS. CosmoDb need initial setup and backend services needs configuration and permissions to read and write to the new database instance.

# Major Changes
- Provide Bicep and bash scripts to provision and configure Azure resources
- Update CICD pipeline to configure backend service to connect to a CosmoDb instance
- Implement sample application client code to access db and define new environment variables
  - COSMOS_ENDPOINT
  - COSMOS_ACCOUNT_NAME
  - COSMOS_DATABASE_NAME
  - COSMOS_MANAGED_IDENTITY (Azure Client Id for the managed identity. NOT the principal id.)
- Refactoring project directory structure, cloud-deployment, moving reusable bicep modules into subfolders
- Update `/api/healthcheck` endpoint to include verifying CosmosDB connectivity and read/write access
- CosmosDb creation/update automated as part of the CICD pipeline. This includes minting a user managed identity with read/write access role to the CosmosDb that can be assigned to a service as an identity.

# Testing/Validation
- Provision CosmoDb instance and managed identity via adhoc Bicep and AZ CLI execution
- Verify CICD pipeline automated configuration of backend service.
- Verify CosmoDb application client connection and read permission from local environment and deployed environment.

# Notes
- For local development, setting COSMOS_MANAGED_IDENTITY is not required. Also, developer will have to execute `az cosmosdb sql role assignment create` with their id and appropriate role definition to read/write to db.
